### PR TITLE
RoslynCodeCompiler  added

### PR DESCRIPTION
**Pull request by alexander-beck**
calling RoslynCodeCompiler via Microsoft.CodeAnalysis.CSharp

solving  https://github.com/mono/t4/issues/68

### CODE REVIEW

**Comment by heikomilke** [comment]
I tested this and I get much faster response times when using TemplateGenerator.ProcessTemplate repeatable from code.

Startup still eats away 500ms but subsequent calls have almost no waiting time

**Comment by mhutch** [comment]
This looks great. I'm not sure about pulling in a dependency of this size by default though, I wonder whether it might make sense to make it opt-in somehow.

**Comment by mhutch** [comment]
I've been thinking about this some more, and I think maybe the best option is to move `RoslynCodeCompiler` to a `Mono.TextTempolating.Roslyn` separate package that adds a `UseInProcessCompiler()` extension method to `TemplatingEngine`. This would register a `Func<CodeCompiler>` callback via an internal method, which (if set) the engine would use to create the compiler.

**Comment by alexander-beck** [comment]
Sounds good. I'll prepare a new PR.

**Comment by mhutch** [comment]
@alexander-beck thanks! I made a few changes, do they look ok to you?

**Comment by alexander-beck** [comment]
Looks great!
I just added the location info to the compiler errors

File: Mono.TextTemplating.Roslyn/RoslynCodeCompiler.cs [hunk 1]
```diff
@@ -73,7 +73,15 @@ public RoslynCodeCompiler (RuntimeInfo runtime)
 				Success = false,
 				Output = new List<string> (),
 				Errors = failures.Select (
-					x => new CodeCompilerError { Message = x.GetMessage () }).ToList (),
+					x => new CodeCompilerError {
+						Message = x.GetMessage(),
+						Column = x.Location.GetMappedLineSpan().StartLinePosition.Character,
```
Comment (mhutch) [inline Mono.TextTemplating.Roslyn/RoslynCodeCompiler.cs]: I would pull this out to a local, it looks like its cost is nontrivial

Comment (alexander-beck) [reply Mono.TextTemplating.Roslyn/RoslynCodeCompiler.cs]: Fixed in #490810f93a90770a3dce0b0776f456e284d2eb1b

Comment (mhutch) [reply Mono.TextTemplating.Roslyn/RoslynCodeCompiler.cs]: Perfect, thank you!

**Pull request merged**